extensions-contrib/influx-extensions/src/main/java/io/druid/data/input/influx/InfluxParser.java

```diff
@@ -0,0 +1,173 @@
+import org.antlr.v4.runtime.TokenStream;
+
+import javax.annotation.Nullable;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class InfluxParser implements Parser<String, Object>
+{
+  public static final String TIMESTAMP_KEY = "__ts";
+  private static final String MEASUREMENT_KEY = "measurement";
+  private final Set<String> measurementWhitelist;
+
+  public InfluxParser(Set<String> measurementWhitelist)
+  {
+    this.measurementWhitelist = measurementWhitelist;
+  }
+
+  @Override
+  public void startFileFromBeginning()
+  {
+  }
+
+  @Nullable
+  @Override
+  public Map<String, Object> parseToMap(String input)
+  {
+    CharStream charStream = new ANTLRInputStream(input);
+    InfluxLineProtocolLexer lexer = new InfluxLineProtocolLexer(charStream);
+    TokenStream tokenStream = new CommonTokenStream(lexer);
+    InfluxLineProtocolParser parser = new InfluxLineProtocolParser(tokenStream);
+
+    List<InfluxLineProtocolParser.LineContext> lines = parser.lines().line();
+    if (parser.getNumberOfSyntaxErrors() != 0) {
+      throw new ParseException("Unable to parse line.");
+    }
+    if (lines.size() != 1) {
+      throw new ParseException("Multiple lines present; unable to parse more than one per record.");
+    }
+
+    Map<String, Object> out = new LinkedHashMap<>();
+
+    InfluxLineProtocolParser.LineContext line = lines.get(0);
+    String measurement = parseIdentifier(line.identifier());
+
+    if (!checkWhitelist(measurement)) {
+      throw new ParseException("Metric not whitelisted.");
+    }
+
+    out.put(MEASUREMENT_KEY, measurement);
+    if (line.tag_set() != null) {
+      line.tag_set().tag_pair().forEach(t -> parseTag(t, out));
+    }
+
+    line.field_set().field_pair().forEach(t -> parseField(t, out));
+
+    if (line.timestamp() != null) {
+      String timestamp = line.timestamp().getText();
+      parseTimestamp(timestamp, out);
+    }
+    return out;
+  }
+
+  private void parseTag(InfluxLineProtocolParser.Tag_pairContext tag, Map<String, Object> out)
+  {
+    String key = parseIdentifier(tag.identifier(0));
+    String value = parseIdentifier(tag.identifier(1));
+    out.put(key, value);
+  }
+
+  private void parseField(InfluxLineProtocolParser.Field_pairContext field, Map<String, Object> out)
+  {
+    String key = parseIdentifier(field.identifier());
+    InfluxLineProtocolParser.Field_valueContext valueContext = field.field_value();
+    Object value;
+    if (valueContext.NUMBER() != null) {
+      value = parseNumber(valueContext.NUMBER().getText());
+    } else if (valueContext.BOOLEAN() != null) {
+      value = parseBool(valueContext.BOOLEAN().getText());
+    } else {
+      value = parseQuotedString(valueContext.QUOTED_STRING().getText());
+    }
+    out.put(key, value);
+  }
+
+  private Object parseQuotedString(String text)
+  {
+    return text.substring(1, text.length() - 1).replaceAll("\\\\\"", "\"");
```

Comment:
Is this the right number of back slashes (5)? Maybe just 4 needed?

Reply:
Yes, 4 slashes is a syntax error. The first four are read into the java string as two backlashes, and the fifth is paired with the double quote, so that the string passed to the regex library is `\\"` which matches the string `\"`, i.e. an escaped quote.